java/ql/src/Security/CWE/CWE-441/UnsafeContentUriResolution.qhelp

```diff
@@ -0,0 +1,49 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>
+  When an Android application wants to access data in a content provider, it uses the <code>ContentResolver</code>
+  object. <code>ContentResolver</code>s communicate with an instance of a class that implements the 
+  <code>ContentProvider</code> interface via URIs with the <code>content://</code> scheme.
+
+  The authority part (the first path segment) of the URI passed as parameter to the <code>ContentResolver</code> 
+  determines which content provider is contacted for the operation. Specific operations that act on files also
+  support the <code>file://</code> scheme, in which case the local filesystem is queried instead.
+
+  If an external component, like a malicious or compromised application, controls the URI that is used in a
+  <code>ContentResolver</code> operation, it can trick the vulnerable application into accessing its own private
+  files or non-exported content providers. Depending on what the vulnerable application does after accessing the file,
+  the attacking application might get access to the file by forcing it to be copied to a public directory like the
+  external storage, or tamper with it by making the application overwrite it with unexpected data.
+</p>
+</overview>
+<recommendation>
+<p>
+  If possible, avoid using externally-provided data to determine URIs used by a <code>ContentResolver</code>.
+  If that is not an option, validate that the incoming URI can only reference trusted components, like an allow list
+  of content providers and/or applications, or alternatively make sure that the URI does not reference private
+  directories like <code>/data/</code>.
+</p>
+</recommendation>
+<example>
+<p>
+  This example shows two ways of opening a file using a <code>ContentResolver</code>. In the first case, externally-provided
+  data coming from an intent is directly used in the file-reading operation, allowing an attacker to provide a URI
+  of the form <code>/data/data/(vulnerable app package)/(private file)</code> to trick the application into reading it and
+  copying it to the external storage. In the second case, the URI is validated before being used, making sure it does not reference
```

Comment:
It looks as if the example contains three cases, two `BAD` and one `GOOD`, so this might need some adjustment.